core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java

```diff
@@ -130,86 +160,76 @@ public Object apply(ServiceRequestContext ctx, HttpRequest req) {
         try (SafeCloseable ignored = RequestContext.push(ctx, false)) {
             return method.invoke(object, parameterValues(ctx, req, parameters, message, requestConverters));
         } catch (Throwable cause) {
-            final HttpResponse response = getResponseForCause(ctx, req, cause);
+            final HttpResponse response = convertException(ctx, req, cause);
             if (response != null) {
                 return response;
             }
             return Exceptions.throwUnsafely(cause);
         }
     }
 
+    /**
+     * Converts the specified {@code result} to {@link HttpResponse}.
+     */
+    private HttpResponse convertResponse(ServiceRequestContext ctx, Object result) {
+        if (result instanceof HttpResponse) {
+            return (HttpResponse) result;
+        }
+        if (result instanceof AggregatedHttpMessage) {
+            return HttpResponse.of((AggregatedHttpMessage) result);
+        }
+
+        for (final ResponseConverterFunction func : responseConverters) {
+            if (func.canConvertResponse(result)) {
+                try (SafeCloseable ignored = RequestContext.push(ctx, false)) {
+                    return func.convertResponse(result);
+                } catch (Exception e) {
+                    throw new IllegalStateException(
+                            "Response converter " + func.getClass().getName() +
+                            " cannot convert a result to HttpResponse: " + result, e);
+                }
+            }
+        }
+        throw new IllegalStateException(
+                "No response converter exists for a result: " +
+                result != null ? result.getClass().getSimpleName() : "(null)");
+    }
+
     /**
      * Returns a {@link HttpResponse} which is created by {@link ExceptionHandlerFunction}.
      */
-    private HttpResponse getResponseForCause(ServiceRequestContext ctx, HttpRequest req,
-                                             Throwable cause) {
+    private HttpResponse convertException(ServiceRequestContext ctx, HttpRequest req,
+                                          Throwable cause) {
         final Throwable rootCause = Throwables.getRootCause(cause);
         for (ExceptionHandlerFunction func : exceptionHandlers) {
-            if (func.accept(rootCause)) {
+            if (func.canHandleException(rootCause)) {
                 try {
-                    return func.handle(ctx, req, rootCause);
+                    return func.handleException(ctx, req, rootCause);
                 } catch (Exception e) {
                     logger.warn("Unexpected exception from an exception handler {}:",
                                 func.getClass().getName(), e);
                 }
             }
         }
-        return ExceptionHandlerFunction.DEFAULT.handle(ctx, req, rootCause);
+        return ExceptionHandlerFunction.DEFAULT.handleException(ctx, req, rootCause);
```

Comment:
How about logging the exception here so that user could know the root cause when none of the exceptionHandlers could not catch the exception???????

Reply:
Ok, good idea. I will write warning log on ExceptionHandler.DEFAULT instance.